### ISSUE TITLE
Can we please use type defs & type checker ?

### ISSUE THREAD

**Issue by Gozala** [ipld/js-ipld-stack#6]
I think actual type definitions are really helpful and eliminates all uncertainties (if done well).  

I'm happy to submit a pull request, if such changes would be welcome.

**Comment by mikeal** [comment]
I’m partial to keeping the code “vanilla JavaScript” with a minimal amount of tooling.

That said, if type defs are in a separate file and help people with particular editors and IDE’s then I would accept that change, I just don’t want the code itself to get complicated by the type defs.

**Comment by Gozala** [comment]
@mikeal what about [comment syntax](https://flow.org/en/docs/types/comments/) ? It let's engines treat code as is and allows typechecker to actually do it's work.

**Comment by mikeal** [comment]
this would add a lot of flow specific comments into the code base. i’ve seen this in other projects and am always thrown off by it, it ads a fair amount of mental overhead, and it also implies that we’re doing type checking when we actually aren’t unless we incorporate the flow toolchain into the project which I definitely don’t want to do.

most developers aren’t using flow and this implies that there’s another set of practices and tools to learn in order to interact with the code base. this isn’t a problem when it’s in a separate file.

**Comment by Gozala** [comment]
> this would add a lot of flow specific comments into the code base. i’ve seen this in other projects and am always thrown off by it, it ads a fair amount of mental overhead, and it also implies that we’re doing type checking when we actually aren’t unless we incorporate the flow toolchain into the project which I definitely don’t want to do.

Yeah I sort of did implied that we'd run flow on changes to ensure that implementation actually complies with types otherwise they're not very useful & can easily diverge.

@vmx in other thread indicated desire to do exactly that & this seemed like a good place to try it out.

> most developers aren’t using flow and this implies that there’s another set of practices and tools to learn in order to interact with the code base. this isn’t a problem when it’s in a separate file.

I'm not suggesting they will have to.

**Comment by Gozala** [comment]
I'll start with a separate file for now.

**Comment by vmx** [comment]
I'm in favour of the Flow inline syntax. I don't use an IDE, but it's good to see the types while coding. Also from my experience, thinking about types makes an API better. If things are hard to be expressed with types, it's likely that the API could be simpler.

**Comment by mikeal** [comment]
My view on type systems in JS is pretty simple, and based on a lot of pain we’ve had in the Node.js ecosystem.

Whenever you create tooling and patterns to paper over deficiencies in the language, you’re signing up for a lot of pain in the future and a lot of additional conceptual overhead for new contributors which makes it harder to contribute and limits the number of new contributors you can reach.

My philosophy is that no programming language is perfect and that it’s always better in the long run to live with and learn to embrace the deficiencies in a language if possible. Being that types are strictly necessary, my view on type systems in JS is that I might use them if they land in the language and can be taken for granted in the platform. Anything we do before that probably isn’t worth it.

I’ve used many type systems in different languages, the first languages I learned were strongly typed, and while there’s some nice things about typing it’s not worth the overhead in JS and it’s better to just live within the constraints of the language as it is today.

**Comment by Gozala** [comment]
You and general JS community already uses linter which is just a weaker version of what flow / ts does. It’s unfortunate you had a bad experience in the past, but I would still suggest giving it a try rather than dismissing based on that other experience.

For what it’s worth I had a pretty good experience with flow because it was helpful in catching many errors that otherwise would have gotten uncaught and influence it had on API design. Also didn’t had any issues with contributors who have never used it before, they just were able to see errors reported by the CI in the pull request and were able to address those without settings anything locally.

If things indeed don’t work out you can always just stop running type-checker / strip out annotations and move on

**Comment by Gozala** [comment]
Type checker also forces you to encode invariants that usually tend to go undocumented (& get broken over time), which as user of lib I find invaluable as it eliminates all the guesswork / source side trips to obtain that knowledge.

**Comment by mikeal** [comment]
> You and general JS community already uses linter which is just a weaker version of what flow / ts does. 

Sure, but linters are widely adopted by a majority of JS developers while flow is not. It’s not the same barrier to entry.

**Comment by Gozala** [comment]
@mikeal You are right - Introducing a type-checking will raise a bar to contributions. I do clam though that: 

1. Bar isn't as high as it may seems & will not prevent motivated contributors who never used flow or ts.
2. It will also raise code & API quality + save significant time on large changes.

I do have a positive experience using flow within Mozilla to support these claims. So all I can offer is give it a chance and judge it yourself.

I also fear my comments on this thread are starting to be more annoying than helpful, so I'll stop & will only respond if explicitly asked.
